### PR TITLE
PLAT-354: Add delegationToken to outgoing messages in SMExecute

### PR DESCRIPTION
**Pull request by dgovorukhin**
**- What I did**
Added delegationToken to outgoing messages in SMExecute

**- How I did it**

Fill field in outgoing messages from field SMExecute

**- How to verify it**

**- Description for the changelog**

https://insolar.atlassian.net/browse/PLAT-354

### CODE REVIEW

**Comment by cyraxred** [comment]
make sure to check - this is the opposite side of adding delegation token
https://github.com/insolar/assured-ledger/pull/319

**Comment by dgovorukhin** [comment]
@cyraxred I am not sure, what I understand your point? Explain please)

File: ledger-core/virtual/authentication/service.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ var deadBeef = [...]byte{0xde, 0xad, 0xbe, 0xef}
 type Service interface {
 	GetCallDelegationToken(outgoing reference.Global, to reference.Global, pn pulse.Number, object reference.Global) payload.CallDelegationToken
 	IsMessageFromVirtualLegitimate(ctx context.Context, payloadObj interface{}, sender reference.Global, pr pulse.Range) (mustReject bool, err error)
+	IsNeedUseToken(token payload.CallDelegationToken) bool
```
Comment (cyraxred) [inline ledger-core/virtual/authentication/service.go:25]: HasToSendToken

File: ledger-core/virtual/execute/execute.go [hunk 2]
```diff
@@ -752,3 +762,9 @@ func NewStateID(pn pulse.Number, data []byte) reference.Local {
 	hash := hasher.Hash(data)
 	return reference.NewLocal(pn, 0, reference.BytesToLocalHash(hash))
 }
+
+func (s *SMExecute) SetToken() {
```
Comment (cyraxred) [inline ledger-core/virtual/execute/execute.go:766]: SetTokenToOutgoing

Comment (bigbes) [inline ledger-core/virtual/execute/execute.go:766]:
```suggestion
func (s *SMExecute) getToken() payload.CallDelegationToken {
    if s.authenticationService.IsTokenIgnored(s.delegationTokenSpec) {
        return payload.CallDelegationToken{}
    }
    return s.delegationTokenSpec
```

File: ledger-core/virtual/execute/execute.go [hunk 3]
```diff
@@ -752,3 +762,9 @@ func NewStateID(pn pulse.Number, data []byte) reference.Local {
...
+	if s.authenticationService.IsNeedUseToken(s.delegationTokenSpec) {
+		s.outgoing.DelegationSpec = s.delegationTokenSpec
+	}
+}
```
Comment (bigbes) [inline ledger-core/virtual/execute/execute.go:770]:
```suggestion
func (s *SMExecute) getToken() payload.CallDelegationToken {
    if s.authenticationService.IsTokenIgnored(s.delegationTokenSpec) {
        return payload.CallDelegationToken{}
    }
    return s.delegationTokenSpec
}
```

File: ledger-core/virtual/execute/execute.go [hunk 4]
```diff
@@ -495,10 +498,12 @@ func (s *SMExecute) stepExecuteOutgoing(ctx smachine.ExecutionContext) smachine.
 		s.outgoing = outgoing.ConstructVCallRequest(s.execution)
 		s.outgoing.CallOutgoing = gen.UniqueIDWithPulse(pulseNumber)
 		s.outgoingObject = reference.NewSelf(s.outgoing.CallOutgoing)
+		s.SetToken()
```
Comment (bigbes) [inline ledger-core/virtual/execute/execute.go:501]:
```suggestion
		s.outgoing.DelegationSpec = s.getToken()
```

File: ledger-core/virtual/execute/execute.go [hunk 5]
```diff
@@ -495,10 +498,12 @@ func (s *SMExecute) stepExecuteOutgoing(ctx smachine.ExecutionContext) smachine.
...
 	case executionevent.CallMethod:
 		s.outgoing = outgoing.ConstructVCallRequest(s.execution)
 		s.outgoing.CallOutgoing = gen.UniqueIDWithPulse(pulseNumber)
 		s.outgoingObject = s.outgoing.Callee
+		s.SetToken()
```
Comment (bigbes) [inline ledger-core/virtual/execute/execute.go:506]:
```suggestion
		s.outgoing.DelegationSpec = s.getToken()
```

File: ledger-core/virtual/execute/execute.go [hunk 6]
```diff
@@ -709,6 +714,11 @@ func (s *SMExecute) stepSendCallResult(ctx smachine.ExecutionContext) smachine.S
 		EntryHeadHash:      nil,
 		ReturnArguments:    executionResult,
 	}
+
+	if s.authenticationService.IsNeedUseToken(s.delegationTokenSpec) {
+		msg.DelegationSpec = s.delegationTokenSpec
+	}
```
Comment (bigbes) [inline ledger-core/virtual/execute/execute.go:720]:
```suggestion
		msg.DelegationSpec = s.getToken()
```

File: ledger-core/virtual/authentication/service.go (hunk 1 above)
Comment (dgovorukhin) [reply ledger-core/virtual/authentication/service.go]: fixed

File: ledger-core/virtual/execute/execute.go (hunk 3 above)
Comment (dgovorukhin) [reply ledger-core/virtual/execute/execute.go]: fixed

File: ledger-core/virtual/execute/execute.go (hunk 4 above)
Comment (dgovorukhin) [reply ledger-core/virtual/execute/execute.go]: fixed

File: ledger-core/virtual/execute/execute.go (hunk 5 above)
Comment (dgovorukhin) [reply ledger-core/virtual/execute/execute.go]: fixed

File: ledger-core/virtual/execute/execute.go (hunk 6 above)
Comment (dgovorukhin) [reply ledger-core/virtual/execute/execute.go]: fixed

File: ledger-core/insolar/payload/marker_auth.go [hunk 7]
```diff
@@ -52,3 +52,9 @@ func (m *VDelegatedCallRequest) customSubject() (Reference, authSubjectMode) {
 func (m *VDelegatedCallResponse) customSubject() (Reference, authSubjectMode) {
 	return m.GetCallee(), hasAuthSubject
 }
+
+// It returns hasAuthSubject ( i.e. for current pulse ) since VDelegatedRequestFinished can come only with
```
Comment (cyraxred) [inline ledger-core/insolar/payload/marker_auth.go:56]: To be "go doc" it should start with method name
// customSubject returns...

Comment (dgovorukhin) [reply ledger-core/insolar/payload/marker_auth.go]: Fixed

**Pull request merged**